### PR TITLE
Fixture.setUp should no longer be overridden.

### PR DESCRIPTION
**Pull request by rbtcollins**
Instead override _setup. (Robert Collins, #1456361, #1456353)

### CODE REVIEW

**Comment by krisztianfekete** [comment]
This is a massive backward incompatible change to the public API, as so far every user has been explicitly advised to do otherwise first:

```
Creating Fixtures
=================
 
Minimally, subclass Fixture, define setUp to initialize your state and schedule
a cleanup for when cleanUp is called and you're done::
```

It is from current doc (1.2.0).

**Comment by rbtcollins** [comment]
Its not backwards incompatible. If you examine the diff you'll see there is a test explicitly testing that the 1.2.0 recommendation still works.

**Comment by krisztianfekete** [comment]
Indeed `test_setUp_subclassed` is there.

Yet, I feel all my test code is immediately in the deprecated category with code against recommendation.
I feel this is a too big price for what we gain:

I expect testing tools
- to provide feedback on failure (to know that something gone wrong) 
- to provide as much details on the failure as possible (to diagnose what went wrong)

the rest is best effort (including isolation of the test environment).

I think this change is in the best effort category.

I liked the original fix by @ZZelle in https://github.com/testing-cabal/testtools/pull/142 as it was improving upon the situation (best effort) while not changing the API or the promises.

----

Overriding `setUp` is natural way of defining a fixture coming from `unittest`, where `TestCase` class is basically a fixture and one or more tests mixed (`testtools` and `fixtures` are a worthy upgrade nicely separating the two concepts).

**Comment by ZZelle** [comment]
useFixture bugs[1] have been created to avoid different  behavior for testtools/fixtures addCleanup. This change solves them by introducing a new difference: unittest/testtools setUp VS fixtures _setUp

Moreover i would prefer this change raises a warning if setUp is overloaded (perhaps i miss it) as it's natural to overload setUp (unittest behavior) not _setUp: fixture developers could/should update their code (as the change does for fixtures in fixtures package).


[1] https://bugs.launchpad.net/testtools/+bug/1456370
[1] https://bugs.launchpad.net/python-fixtures/+bug/1456353
[2] currently testtools calls cleanups even if setUp fails, fixtures only if setUp succeed

**Comment by rbtcollins** [comment]
The bugs were only somewhat about differing behaviour: they were primarily about leaking global state due to writing hermetic setUp's in fixtures being hard. Even if framed as being primarily a difference thing - unittest tearDown doesn't run if setUp fails, thats one reason why cleanups do trigger on TestCase - we wanted to fix things backward compatibly, and most / all tearDown methods as written wouldn't operate correctly if run from a failed setUp: the APIs are just too awkward. clean-up style code is a lot easier to get right. testtools.TestCase can call its own cleanUps because it knows the internal state of the TestCase; it can't know the internal state of a fixture in the same way.

Further, the fundamental driver for addressing it this way is with: syntax. There are two symmetries here:
* fixtures are context managers; setUp == __enter__
* fixtures have a similar API to TestCase; f.setUp is like t.setUp
The former one is foundational: its what makes Fixtures nice to work with. The second is about discoverability and familiarity. They are both important, but not entirely compatible, because TestCase's are not context managers. Thats why I believe the context manager symmetry needs to win here IMO.

A warning on setUp being overloaded could be interesting. On the one hand it makes folk like @krisztianfekete with lots of fixtures feel even more sad, because their code will instantly start whinging at them about needing to change. On the other hand it will alert people that they need to do something. Introducing a warning like that in the release that offers a better way is often the short path to pissing users off in my experience: because they have to choose between the new features and being shouted at. Waiting a few releases so that they're *already using* the latest code and then introducing a deprecation like that makes more sense to me.

I think I should alter the documentation in the patch to make it more inclusive: make it clear that overriding setUp *is* supported, what you need to do if you override setUp (that is be atomic or not). My intent was to offer a better way, not to stop supporting existing code. I'll prep that and push it up into this PR later today.

File: NEWS [hunk 1]
```diff
@@ -9,6 +9,9 @@ NEXT
 * Add MockPatch, MockPatchMultiple, MockPatchObject - adapters to mock.
   (Julien Danjou, Robert Collins)
 
+* Fixture.setUp should no longer be overridden in subclasses. Instead
+  override _setup. (Robert Collins, #1456361, #1456353)
```
Comment (jml) [inline NEWS]: Please spell `_setUp` consistently. (But I think you're revising this anyway)

File: README [hunk 2]
```diff
@@ -155,8 +152,8 @@ clean up after a fixture has been used, all fixtures define a ``cleanUp``
 method which should be called when a fixture is finished with.
 
 Because its nice to be able to build a particular set of related fixtures in
```
Comment (jml) [inline README]: s/its/it's/

File: fixtures/fixture.py [hunk 3]
```diff
@@ -49,6 +50,13 @@ def combine_details(source_details, target_details):
         target_details[name] = content_object
 
 
+class SetupError(Exception):
+    """Setup failed.
+
+    args[0] will be a details dict.
+    """
```
Comment (jml) [inline fixtures/fixture.py]: Maybe make `.details` an instance variable as well? Seems like a nicer API.

File: fixtures/fixture.py [hunk 4]
```diff
@@ -153,16 +174,49 @@ def getDetails(self):
...
+
+    def _setUp(self):
+        """Template method for subclasses to override.
+
+        Override this to customise the fixture. When overriding
+        be sure to include self.addCleanup calls to restore the fixture to
+        an un-setUp state, so that a single Fixture instance can be reused.
+
+        :return: None.
+        """
```
Comment (jml) [inline fixtures/fixture.py]: Perhaps this should raise `NotImplementedError`?

File: fixtures/fixture.py [hunk 5]
```diff
@@ -183,15 +237,23 @@ def useFixture(self, fixture):
         """Use another fixture.
 
         The fixture will be setUp, and self.addCleanup(fixture.cleanUp) called.
+        If the fixture fails to setup, useFixture will attempt to gather it's
```
Comment (jml) [inline fixtures/fixture.py]: s/setup/set up/ ("setup" is the noun, "set up" the verb phrase)

File: fixtures/tests/test_fixture.py [hunk 6]
```diff
@@ -185,7 +203,54 @@ def test_addDetail(self):
...
+            self.assertEqual(1, f.fred)
+        self.assertEqual(2, f.fred)
+
+    def test__setUp_fails(self):
+        class Subclass(fixtures.Fixture):
+            def _setUp(self):
+                self.addDetail('log', text_content('stuff'))
+                1/0
+        f = Subclass()
+        e = self.assertRaises(fixtures.MultipleExceptions, f.setUp)
```
Comment (jml) [inline fixtures/tests/test_fixture.py]: In this and the following test, there are a lot of assertions and it takes a chunk of time to parse them. I think it's worth summarizing what these test expect to happen in comments.

File: fixtures/tests/test_fixture.py [hunk 7]
```diff
@@ -162,7 +180,7 @@ def test_details_from_child_fixtures_are_returned(self):
             self.assertEqual({}, parent.getDetails())
             # After cleanup the child details are still gone.
             child.addDetail('foo', 'content')
-        self.assertEqual({}, parent.getDetails())
+        self.assertRaises(TypeError, parent.getDetails)
```
Comment (jml) [inline fixtures/tests/test_fixture.py]: It's not clear to me where the `TypeError` comes from. I'm guessing it's the fixture trying to do something to some element of its internal state after that state has been set to `None`. 

On the one hand, I think this is not the most helpful interface. On the other, resetting to `None` is probably the simplest and cleanest implementation. Thoughts?

File: NEWS (hunk 1 above)
Comment (rbtcollins) [inline NEWS]: Done

File: README (hunk 2 above)
Comment (rbtcollins) [inline README]: Done

File: fixtures/fixture.py (hunk 3 above)
Comment (rbtcollins) [inline fixtures/fixture.py]: I was hesitating to add noise - args[] is the base contract for exceptions. I'll have a fiddle and see what it looks like: we can always add that later.

File: fixtures/fixture.py (hunk 4 above)
Comment (rbtcollins) [inline fixtures/fixture.py]: No, it can't (if it did, the change would be backwards incompatible).

File: fixtures/fixture.py (hunk 5 above)
Comment (rbtcollins) [inline fixtures/fixture.py]: Fixed. Also fixed its.

File: fixtures/tests/test_fixture.py (hunk 6 above)
Comment (rbtcollins) [inline fixtures/tests/test_fixture.py]: Done

File: fixtures/tests/test_fixture.py (hunk 7 above)
Comment (rbtcollins) [inline fixtures/tests/test_fixture.py]: getDetails tries to dereference self._details. So this is just checking things are indeed wiped out, and its not really a user interface.

**Comment by rbtcollins** [comment]
@ZZelle I've revised the commit message as you say.

**Pull request merged**